packages/server-wallet/src/message-service/socket-io-message-service.ts

```diff
@@ -22,13 +30,32 @@ export class SocketIOMessageService implements MessageServiceInterface {
     this.peers.push(socket);
   }
 
+  public setLatencyOptions(options: LatencyOptions): void {
+    this._latencyOptions = options;
+  }
+
   public async send(messages: Message[]): Promise<void> {
-    if (messages.length > 0) {
-      this.server.emit('message', messages);
+    if (this._frozen) {
+      this._messageQueue.push(...messages);
+    } else {
+      const shouldDrop = Math.random() > 1 - this._latencyOptions.dropRate;
+
+      if (!shouldDrop) {
+        const {meanDelay} = this._latencyOptions;
+        if (meanDelay && meanDelay > 0) {
+          const delayAmount = meanDelay / 2 + Math.random() * meanDelay;
+
+          await delay(delayAmount, {signal: this._abortController.signal});
+        }
+        if (!this._destroyed) {
+          this.server.emit('message', messages);
+        }
+      }
     }
   }
 
   public async destroy(): Promise<void> {
+    this._destroyed = true;
```

Comment:
Do we need to call `this._abortController.abort();`?

Reply:
Done in 4d37d37